File: src/server/src/velvet_dawn/units/movement.py
from typing import List, Dict

import velvet_dawn
from velvet_dawn import errors
import velvet_dawn.map.neighbours
from velvet_dawn.config import Config
from velvet_dawn.dao import db
from velvet_dawn.db.instances import UnitInstance
from velvet_dawn.models.phase import Phase
from velvet_dawn.models.coordinate import Coordinate
from velvet_dawn.dao.models import Player, TileInstance


def get_remaining_moves(unit: UnitInstance):
    # TODO Test this when influence is done
    # TODO Add influence here and other mechanics here
    return unit.get_attribute("movement.remaining", default=0)


def move(player: Player, entity_pk: int, path: List[dict], config: Config):
    """ Move an entity through a path

    Args:
        player: The player moving the entity
        entity_pk: The db instance id of the entity
        path: List of {x, y} positions the entity moves, starting with it's current pos
        config: The came config
    """
    if velvet_dawn.game.phase.get_phase() != Phase.GAME:
        raise errors.GamePhaseError("game")

    if velvet_dawn.game.turns.get_active_turn(Phase.GAME) != player.team:
        raise errors.InvalidTurnError()

    entity = velvet_dawn.db.units.get_unit_by_instance_id(entity_pk)
    if entity is None:
        raise errors.ItemNotFoundError(entity_pk)

    if entity.player != player.name:
        raise errors.ValidationError("You may only move your own pieces.")

    # Validate correct path and calculate remaining moves
    remaining_moves = _validate_entity_traversing_path(entity, path, config)

    # Run leave movement triggers
    trigger_on_leave_actions(entity, velvet_dawn.map.get_tile(entity.x, entity.y))

    # Update the entity to the new position based on the path and the remaining moves
    entity.set_attribute("movement.remaining", remaining_moves)
    instance = velvet_dawn.db.units.move(entity, path[-1]['x'], path[-1]['y'])
    db.session.commit()

    # Run enter movement triggers
    trigger_on_enter_actions(
        instance,
        velvet_dawn.map.get_tile(instance.tile_x, instance.tile_y)
    )

    return instance


def _validate_entity_traversing_path(entity: UnitInstance, path: List[Dict[str, int]], config: Config) -> int:
    """ Validate an entity's traversal path

    This function will raise exceptions if the path is invalid.

    Args:
        entity: The entity to move
        path: The list of {x: , y: } point the entity traverses

    Returns:
        Number of remaining moves the entity has

    Raises:
        Exception if the path is invalid
    """
    tiles = [velvet_dawn.map.get_tile(point['x'], point['y']) for point in path]
    remaining_moves = get_remaining_moves(entity)

    if tiles[0].x != entity.x or tiles[0].y != entity.y:
        raise errors.EntityMovementErrorInvalidStartPos()

    for i, tile in enumerate(tiles):
        if not tile:
            raise errors.EntityMovementErrorInvalidItem()

        # Check if the previous tile is a neighbour
        if i > 0 and tiles[i - 1] not in velvet_dawn.map.neighbours.get_neighbours(Coordinate(tile.x, tile.y), config):
            raise errors.EntityMovementErrorNotNeighbours(i + 1)

        # If the player already exists in this tile then we don't need to
        # check if it's valid or decrement the remaining moves
        if tile.x == entity.x and tile.y == entity.y:
            continue

        tile_definition = velvet_dawn.datapacks.tiles.get(tile.tile_id)
        if not tile_definition:
            raise errors.UnknownTile(tile.tile_id)

        if not velvet_dawn.map.is_traversable(tile.x, tile.y):
            raise errors.EntityMovementErrorTileNotTraversable(tile)

        if remaining_moves <= 0:
            raise errors.EntityMovementErrorTileNoRemainingMoves()

        remaining_moves -= velvet_dawn.map.get_tile_movement_weight(tile)

    return remaining_moves


def trigger_on_leave_actions(unit_instance: UnitInstance, tile_instance: TileInstance):
    """ Trigger on leave actions

    This occurs when an entity leaves a tile, both the tile and
    unit will be called here

    This function is testing as part of the trigger testing suite
    not movement suite

    Args:
        unit_instance: The unit to trigger on
        tile_instance: The tile to trigger on
    """
    velvet_dawn.datapacks.entities[unit_instance.entity_id].triggers.on_leave(unit_instance)
    velvet_dawn.datapacks.tiles[tile_instance.tile_id].triggers.on_leave(tile_instance)


def trigger_on_enter_actions(unit_instance: UnitInstance, tile_instance: TileInstance):
    """ Trigger on enter actions

    This occurs when an entity enters a tile, both the tile and
    unit will be called here

    This function is testing as part of the trigger testing suite
    not movement suite

    Args:
        unit_instance: The unit to trigger on
        tile_instance: The tile to trigger on
    """
    velvet_dawn.datapacks.entities[unit_instance.entity_id].triggers.on_enter(unit_instance)
    velvet_dawn.datapacks.tiles[tile_instance.tile_id].triggers.on_enter(tile_instance)

File: src/jserver/src/velvetdawn/entities/Movement.java
package velvetdawn.entities;

import velvetdawn.VelvetDawn;
import velvetdawn.constants.AttributeKeys;
import velvetdawn.models.Coordinate;
import velvetdawn.models.Phase;
import velvetdawn.models.anytype.Any;
import velvetdawn.models.instances.TileInstance;
import velvetdawn.models.instances.entities.EntityInstance;

import java.util.List;
import java.util.stream.Collectors;

public class Movement {

    private final VelvetDawn velvetDawn;

    public Movement(VelvetDawn velvetDawn) {
        this.velvetDawn = velvetDawn;
    }

    public int getRemainingMoves(EntityInstance entity) {
        return (int) entity.attributes.get(AttributeKeys.EntityRemainingMoves, Any.from(0)).toNumber();
    }

    /** Move an entity along a path
     * 
     * @param entity The entity to move
     * @param path The path to move the entity along
     */
    public EntityInstance move(EntityInstance entity, List<Coordinate> path) throws Exception {
        if (velvetDawn.game.phase != Phase.Game)
            throw new Exception("Invalid game phase to move entities");
    
        if (velvetDawn.game.turns.getActiveTurn() != entity.player.team)
            throw new Exception("You may only move entities on your turn");
    
        // Validate correct path and calculate remaining moves
        var remainingMoves = validateEntityTraversingPath(entity, path);
    
        // Run leave movement triggers
        triggerOnLeaveActions(entity, velvetDawn.map.getTile(entity.position));
    
        // Update the entity to the new position based on the path and the remaining moves
        entity.attributes.set("movement.remaining", Any.from(remainingMoves));
        velvetDawn.entities.setPosition(entity, path.get(path.size() - 1));

        // Run enter movement triggers
        triggerOnEnterActions(entity, velvetDawn.map.getTile(entity.position));
    
        return entity;
    }

    /** Validate an entity's traversal path
     * 
     * This function will raise an exception if the path is invalid.
     * 
     * @param entity The entity to move
     * @param path The path to move the entity
     */
    public int validateEntityTraversingPath(EntityInstance entity, List<Coordinate> path) throws Exception {
        var tiles = path.stream().map(velvetDawn.map::getTile).collect(Collectors.toList());
        var remainingMoves = getRemainingMoves(entity);

        if (!tiles.get(0).position.equals(entity.position))
            throw new Exception("Invalid start position of the entity. The first position in the list, should be where the entity exists.");

        for (int i = 0; i < tiles.size(); i++) {
            var tile = tiles.get(i);

            if (tile == null)
                throw new Exception("null tile found path");

            // Check if the previous tile is a neighbour
            if (i > 0) {
                var prevTile = tiles.get(i - 1);
                if (velvetDawn.map.getNeighbours(tile.position).stream().noneMatch(x -> x.equals(prevTile.position)))
                    throw new Exception("Not all items in the path are neighbours to one-another");
            }

            // If the player already exists in this tile then we don't need to
            // check if it's valid or decrement the remaining moves
            if (tile.position.equals(entity.position))
                continue;

            var tileDefinition = velvetDawn.datapacks.tiles.get(tile.datapackId);
            if (tileDefinition == null)
                throw new Exception(String.format("Unknown tile '%s'", tile.datapackId));

            if (!velvetDawn.map.isTraversable(tile.position))
                throw new Exception("Path traverses through a non-traversable tile");

            if (remainingMoves <= 0)
                throw new Exception("Entity has no remaining moves");

            remainingMoves -= tile.attributes.get(AttributeKeys.TileMovementWeight).toNumber();
        }

        return remainingMoves;
    }
    
    /** Trigger on leave actions
     * 
     * This occurs when an entity leaves a tile, both the tile and unit will be
     * passed here. This function is testing as part of the trigger testing 
     * suite, not movement suite.
     * 
     * @param entity This unit to trigger on
     * @param tile The tile to trigger on
     */
    public void triggerOnLeaveActions(EntityInstance entity, TileInstance tile) throws Exception {
        velvetDawn.datapacks.entities.get(entity.datapackId).triggers.onLeave(entity);
        velvetDawn.datapacks.tiles.get(tile.datapackId).triggers.onLeave(tile);
    }
                
    /** Trigger on enter actions
     * 
     * This occurs when an entity enters a tile, both the tile and unit will be
     * called here.
     * 
     * This function is testing as part of the trigger testing suite, not 
     * movement quite.
     * 
     * @param entity The entity to trigger move actions for
     * @param tileInstance The tile the unit moved to.
     */
    public void triggerOnEnterActions(EntityInstance entity, TileInstance tileInstance) throws Exception {
        velvetDawn.datapacks.entities.get(entity.datapackId).triggers.onEnter(entity);
        velvetDawn.datapacks.tiles.get(tileInstance.datapackId).triggers.onEnter(tileInstance);
    }
}



